pkg/controller/common/stackmon/validations/validations.go

```diff
@@ -19,6 +19,7 @@ const (
 	invalidElasticsearchRefsMsg = "Only one Elasticsearch reference is supported for %s Stack Monitoring"
 
 	InvalidKibanaElasticsearchRefForStackMonitoringMsg = "Kibana must be associated to an Elasticsearch cluster through elasticsearchRef in order to enable monitoring metrics features"
+	InvalidBeatsElasticsearchRefForStackMonitoringMsg  = "Beats must be associated to an Elasticsearch cluster through elasticsearchRef in order to enable monitoring metrics features"
```

Comment:
I think we would want an actual validation in the webhook/reconciler using this (in addition to the error handling you implemented during config generation)

Reply:
Ok, additional validation added to the default webhook checks for Beat that validates the monitoring section of the specificiation.  *Note* that these are slightly different than the `common.stackmon.validations`, in that they break the "refs defined", and the length of the refs into 2 separate distinct validations+errors.

https://github.com/elastic/cloud-on-k8s/blob/main/pkg/controller/common/stackmon/validations/validations.go#L44

This doesn't seem right, as if the refs are not defined, you'll never get a validation error:
```
if monitoring.AreEsRefsDefined(refs) && len(refs) != 1 {
```

Should we create an issue, and get the stackmon validations updated as well?

Reply:
Not sure I fully understand. At the top of the common validations there is a `monitoring.IsDefined(resource)` call which is a logical OR or logs or metrics refs defined. So we know at least one is defined, otherwise we would exit the whole validation function. After that it is just a matter of which one or if both are defined and if they are defined correctly.

Reply:
I hope this makes it more clear.  The following block only checks if the version is not supported, and only returns early if one is defined AND the version is not supported:
```
if monitoring.IsDefined(resource) {
		err := IsSupportedVersion(version)
		if err != nil {
			finalMinStackVersion, _ := semver.FinalizeVersion(MinStackVersion.String()) // discards prerelease suffix
			errs = append(errs, field.Invalid(field.NewPath("spec").Child("version"), version,
				fmt.Sprintf(UnsupportedVersionMsg, finalMinStackVersion)))
		}
	}
```

`IsDefined` checks the following, so if the monitoring definition just has `Namespace`, then it won't be technically `Defined`.
```
// IsDefined checks if the object selector is not nil and has a name or a secret name.
// Namespace is not mandatory as it may be inherited by the parent object.
func (o *ObjectSelector) IsDefined() bool {
	return o != nil && o.NameOrSecretName() != ""
}
```

Let's say someone does this
```
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
spec:
  version: 8.2.3
  monitoring:
    metrics:
      elasticsearchRefs:
      - namespace: "test",
```

So the `IsSupportedVersion` check will be skipped, as this isn't `Defined` and won't return early, and the following check will be skipped `if monitoring.AreEsRefsDefined(refs) && len(refs) != 1 {` as this monitoring ref isn't `Defined` as well, and the subsequent logs check will be skipped as well, as the logs block isn't defined at all, and this `Validate` function will return no errors, although this clearly isn't a valid monitoring block...

Reply:
Forgot to mention, I check this exact situation [here](https://github.com/elastic/cloud-on-k8s/pull/5878/files#diff-e391f05ef44e6228796152fb00ef6b2ff0c2766f2d330bb0e4e979b759b951e1R258) for beats monitoring

Reply:
PR Updated to use common validation